jersey-servers/src/main/java/com/palantir/remoting2/servers/jersey/CompressionFilter.java

```diff
@@ -0,0 +1,160 @@
+
+    public CompressionFilter() {
+        this(0);
+    }
+
+    public CompressionFilter(int minCompressionBytes) {
+        this.minCompressionBytes = minCompressionBytes;
+        this.parsedHeaders = CacheBuilder.newBuilder()
+                .maximumSize(100)
+                .build(new CacheLoader<List<String>, List<String>>() {
+                    @Override
+                    public List<String> load(List<String> headerLines) throws Exception {
+                        return acceptedEncodings(headerLines);
+                    }
+                });
+    }
+
+    @Override
+    public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext)
+            throws IOException {
+        if (isTooSmall(responseContext) || isEncoded(responseContext) || isUncompressable(responseContext)) {
+            return;
+        }
+
+        List<String> headerLines = requestContext.getHeaders().get(HttpHeaders.ACCEPT_ENCODING);
+        if (headerLines == null) {
+            return;
+        }
+
+        List<String> acceptedEncodings = parsedHeaders.getUnchecked(headerLines);
+        for (Encoding encoding : Encoding.values()) {
+            if (acceptedEncodings.contains(encoding.encoding)) {
+                MultivaluedMap<String, Object> headers = responseContext.getHeaders();
+
+                // Set 'Vary: Accept-Encoding' so intermediate caches differentiate between differently
+                // encoded responses
+                headers.add(HttpHeaders.VARY, HttpHeaders.ACCEPT_ENCODING);
+                headers.add(HttpHeaders.CONTENT_ENCODING, encoding.encoding);
+                responseContext.setEntityStream(encoding.compressor.apply(responseContext.getEntityStream()));
+                return;
+            }
+        }
+    }
+
+    private boolean isTooSmall(ContainerResponseContext response) {
+        int length = response.getLength();
+        return length != -1 && length < minCompressionBytes;
+    }
+
+    private static boolean isEncoded(ContainerResponseContext response) {
+        List<Object> encodings = response.getHeaders().get(HttpHeaders.CONTENT_ENCODING);
+        return encodings != null && encodings.size() > 0;
+    }
+
+    private static boolean isUncompressable(ContainerResponseContext response) {
+        MediaType mediaType = MediaType.parse(response.getMediaType().toString());
+        return UNCOMPRESSIBLE.stream().anyMatch(mediaType::is);
+    }
+
+    private static List<String> acceptedEncodings(List<String> headerLines) {
+        ImmutableList.Builder<String> builder = ImmutableList.builder();
+
+        // Single line looks like `deflate, gzip;q=1.0, *;q=0.5`
+        // See section 14.3: https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html
+        // TODO(jellis): don't ignore qvalues
+        for (String line : headerLines) {
+            for (String acceptedEncoding : line.split(",")) {
+                builder.add(acceptedEncoding.split(";")[0]);
```

Comment:
Maybe I'm missing something obvious here, but shouldn't we be trimming whitespace based on the example in the comment above? In Chrome, my Accept-Encodings is `gzip, deflate, sdch, br`, which results in acceptedEncodings of `["gzip", " deflate", " sdch", " br"]`.